api/src/main/java/org/apache/iceberg/util/DataFileSet.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.util;
+
+import java.util.Objects;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+
+public class DataFileSet extends WrapperSet<DataFile> {
+  private static final ThreadLocal<DataFileWrapper> WRAPPERS =
+      ThreadLocal.withInitial(() -> DataFileWrapper.wrap(null));
+
+  private DataFileSet() {
+    // needed for serialization/deserialization
+  }
+
+  private DataFileSet(Iterable<Wrapper<DataFile>> wrappers) {
+    super(wrappers);
+  }
+
+  public static DataFileSet create() {
+    return new DataFileSet();
+  }
+
+  public static DataFileSet of(Iterable<? extends DataFile> iterable) {
+    return new DataFileSet(
+        Iterables.transform(
+            iterable,
+            obj -> {
+              Preconditions.checkNotNull(obj, "Invalid object: null");
```

Comment:
We could technically move the precondition into `wrap` and get rid of most preconditions in the parent class.

```
private static final ThreadLocal<DataFileWrapper> WRAPPERS =
    ThreadLocal.withInitial(DataFileWrapper::new);

...

private static class DataFileWrapper implements Wrapper<DataFile> {
  private DataFile file;

  private DataFileWrapper() {}

  private static DataFileWrapper wrap(DataFile dataFile) {
    Preconditions.checkNotNull(dataFile, "Invalid object: null");
    DataFileWrapper wrapper = new DataFileWrapper();
    wrapper.set(dataFile);
    return wrapper;
  }

...

}

```

Either way works for me. Let's get this in!

Reply:
I would prefer to keep the precondition checks in the Set implementation itself, since it reflects the contract of the Set itself. Also I think the goal is to keep wrapper instances to a minimum (hence the ` wrapper.set(null)` calls after usage). Let's get this in and we can always revisit and change this later if we decide there's a better way to handle this. Thanks for the reviews